File: src/mcp/service_config.py
"""
MCP Gateway Service Configuration

Service configuration dataclass and management functions for MCP services.
"""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional


@dataclass
class McpServiceConfig:
    """MCP service configuration."""
    service_id: str
    name: str
    path: str
    enabled: bool = True
    builtin: bool = False
    health_url: Optional[str] = None
    transport: str = "streamable-http"


# MCP service configuration registry
_MCP_SERVICE_CONFIGS: Dict[str, McpServiceConfig] = {
    "memory": McpServiceConfig(service_id="memory", name="Memory", path="/memory", enabled=True, builtin=True),
    "graph": McpServiceConfig(service_id="graph", name="Graph", path="/graph", enabled=True, builtin=True),
    "search": McpServiceConfig(service_id="search", name="Search", path="/search", enabled=True, builtin=True),
    "workflow": McpServiceConfig(service_id="workflow", name="Workflow", path="/workflow", enabled=True, builtin=True),
    "critic": McpServiceConfig(service_id="critic", name="Critic", path="/critic", enabled=True, builtin=True),
    "agent": McpServiceConfig(service_id="agent", name="Agent", path="/agent", enabled=True, builtin=True),
    "skills": McpServiceConfig(service_id="skills", name="Skills", path="/skills", enabled=True, builtin=True),
}

# Service health status cache
_MCP_SERVICE_HEALTH_CACHE: Dict[str, str] = {service_id: "unknown" for service_id in _MCP_SERVICE_CONFIGS}

# Runtime server order for consistent display
_RUNTIME_SERVER_ORDER = ["memory", "graph", "search", "workflow", "critic", "agent", "skills"]


def _serialize_service_config(config: McpServiceConfig, services: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
    """Serialize service config to JSON-serializable dict."""
    runtime_status = "unknown"
    if services is not None:
        runtime_status = _service_runtime_status(config.service_id, services)
    elif config.service_id in _MCP_SERVICE_HEALTH_CACHE:
        runtime_status = _MCP_SERVICE_HEALTH_CACHE[config.service_id]

    return {
        "id": config.service_id,
        "name": config.name,
        "path": config.path,
        "enabled": config.enabled,
        "builtin": config.builtin,
        "transport": config.transport,
        "health_url": config.health_url,
        "status": runtime_status,
    }


def _normalize_service_config_payload(service_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
    """Normalize and validate service config payload."""
    normalized = {
        "service_id": service_id.strip().lower(),
        "name": str(body.get("name") or service_id).strip(),
        "path": str(body.get("path") or f"/{service_id}").strip(),
        "enabled": bool(body.get("enabled", True)),
        "builtin": bool(body.get("builtin", False)),
        "health_url": body.get("health_url"),
        "transport": str(body.get("transport") or "streamable-http").strip() or "streamable-http",
    }

    if not normalized["service_id"]:
        raise ValueError("service_id is required")

    if not normalized["path"]:
        raise ValueError("path is required")

    if not normalized["path"].startswith("/"):
        normalized["path"] = f"/{normalized['path']}"

    if normalized["health_url"] is not None:
        normalized["health_url"] = str(normalized["health_url"]).strip() or None

    return normalized


def _update_service_config(service_id: str, body: Dict[str, Any], *, create_if_missing: bool = False) -> McpServiceConfig:
    """Update or create service configuration."""
    current = _MCP_SERVICE_CONFIGS.get(service_id)
    if current is None and not create_if_missing:
        raise KeyError(service_id)

    if current is not None:
        if current.builtin and body.get("path") is not None:
            raise ValueError("builtin service path is immutable")
        if current.builtin and body.get("builtin") is False:
            raise ValueError("builtin service cannot be downgraded")

    payload = _normalize_service_config_payload(service_id, body)

    if current is None:
        updated = McpServiceConfig(**payload)
    else:
        updated = McpServiceConfig(
            service_id=current.service_id,
            name=payload["name"],
            path=current.path if current.builtin else payload["path"],
            enabled=payload["enabled"],
            builtin=current.builtin,
            health_url=payload["health_url"],
            transport=payload["transport"],
        )

    _MCP_SERVICE_CONFIGS[updated.service_id] = updated
    _MCP_SERVICE_HEALTH_CACHE.setdefault(updated.service_id, "unknown")
    return updated


def _set_service_enabled(service_id: str, enabled: bool) -> McpServiceConfig:
    """Enable or disable a service."""
    config = _MCP_SERVICE_CONFIGS.get(service_id)
    if config is None:
        raise KeyError(service_id)

    if config.builtin and not enabled:
        raise ValueError("builtin service cannot be disabled")

    updated = McpServiceConfig(
        service_id=config.service_id,
        name=config.name,
        path=config.path,
        enabled=enabled,
        builtin=config.builtin,
        health_url=config.health_url,
        transport=config.transport,
    )
    _MCP_SERVICE_CONFIGS[service_id] = updated
    return updated


def _refresh_service_health_cache(services: Dict[str, str]) -> None:
    """Refresh the service health status cache."""
    for service_id in _MCP_SERVICE_CONFIGS:
        _MCP_SERVICE_HEALTH_CACHE[service_id] = _service_runtime_status(service_id, services)


def _service_runtime_status(service_id: str, services: Dict[str, str]) -> str:
    """Get service runtime status from health check results."""
    config = _MCP_SERVICE_CONFIGS.get(service_id)
    if config and not config.enabled:
        return "disabled"
    return services.get(service_id, "unknown")


__all__ = [
    "McpServiceConfig",
    "_MCP_SERVICE_CONFIGS",
    "_MCP_SERVICE_HEALTH_CACHE",
    "_RUNTIME_SERVER_ORDER",
    "_serialize_service_config",
    "_normalize_service_config_payload",
    "_update_service_config",
    "_set_service_enabled",
    "_refresh_service_health_cache",
    "_service_runtime_status",
]

File: src-ts/mcp/service-config.ts
/**
 * MCP Gateway Service Configuration
 *
 * Service configuration registry and management functions for MCP services.
 *
 * Migrated from src/mcp/service_config.py
 */

// ============================================================
// McpServiceConfig Interface
// ============================================================

export type McpTransport = 'streamable-http' | 'stdio' | 'sse';

const VALID_TRANSPORTS: readonly McpTransport[] = ['streamable-http', 'stdio', 'sse'];

export function validateMcpTransport(value: string): McpTransport {
  if (VALID_TRANSPORTS.includes(value as McpTransport)) return value as McpTransport;
  throw new Error(`Invalid transport "${value}". Must be one of: ${VALID_TRANSPORTS.join(', ')}`);
}

export interface McpServiceConfig {
  serviceId: string;
  name: string;
  path: string;
  enabled: boolean;
  builtin: boolean;
  healthUrl: string | null;
  transport: McpTransport;
}

// ============================================================
// Service Configuration Registry
// ============================================================

export const MCP_SERVICE_CONFIGS: Record<string, McpServiceConfig> = {
  memory: {
    serviceId: 'memory',
    name: 'Memory',
    path: '/memory',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
  graph: {
    serviceId: 'graph',
    name: 'Graph',
    path: '/graph',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
  search: {
    serviceId: 'search',
    name: 'Search',
    path: '/search',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
  workflow: {
    serviceId: 'workflow',
    name: 'Workflow',
    path: '/workflow',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
  critic: {
    serviceId: 'critic',
    name: 'Critic',
    path: '/critic',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
  agent: {
    serviceId: 'agent',
    name: 'Agent',
    path: '/agent',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
  skills: {
    serviceId: 'skills',
    name: 'Skills',
    path: '/skills',
    enabled: true,
    builtin: true,
    healthUrl: null,
    transport: 'streamable-http',
  },
};

/** Service health status cache. */
export const MCP_SERVICE_HEALTH_CACHE: Record<string, string> = Object.fromEntries(
  Object.keys(MCP_SERVICE_CONFIGS).map((id) => [id, 'unknown']),
);

/** Runtime server order for consistent display. */
export const RUNTIME_SERVER_ORDER: readonly string[] = [
  'memory', 'graph', 'search', 'workflow', 'critic', 'agent', 'skills',
] as const;

// ============================================================
// Public Functions
// ============================================================

/**
 * Get service runtime status from health check results.
 */
export function serviceRuntimeStatus(
  serviceId: string,
  services: Record<string, string>,
): string {
  const config = MCP_SERVICE_CONFIGS[serviceId];
  if (config && !config.enabled) return 'disabled';
  return services[serviceId] ?? 'unknown';
}

/**
 * Serialize service config to JSON-serializable dict.
 */
export function serializeServiceConfig(
  config: McpServiceConfig,
  services?: Record<string, string>,
): Record<string, unknown> {
  let runtimeStatus = 'unknown';
  if (services !== undefined) {
    runtimeStatus = serviceRuntimeStatus(config.serviceId, services);
  } else if (config.serviceId in MCP_SERVICE_HEALTH_CACHE) {
    runtimeStatus = MCP_SERVICE_HEALTH_CACHE[config.serviceId];
  }

  return {
    id: config.serviceId,
    name: config.name,
    path: config.path,
    enabled: config.enabled,
    builtin: config.builtin,
    transport: config.transport,
    health_url: config.healthUrl,
    status: runtimeStatus,
  };
}

/**
 * Normalize and validate service config payload.
 */
export function normalizeServiceConfigPayload(
  serviceId: string,
  body: Record<string, unknown>,
): McpServiceConfig {
  const transport = validateMcpTransport(String(body.transport ?? 'streamable-http').trim() || 'streamable-http');
  let serviceIdValue = serviceId.trim().toLowerCase();
  let nameValue = String(body.name ?? serviceId).trim();
  let pathValue = String(body.path ?? `/${serviceId}`).trim();
  const enabledValue = Boolean(body.enabled ?? true);
  const builtinValue = Boolean(body.builtin ?? false);
  const healthUrlValue = body.health_url ?? null;

  if (!serviceIdValue) {
    throw new Error('service_id is required');
  }
  if (!pathValue) {
    throw new Error('path is required');
  }
  if (!pathValue.startsWith('/')) {
    pathValue = `/${pathValue}`;
  }

  return {
    serviceId: serviceIdValue,
    name: nameValue,
    path: pathValue,
    enabled: enabledValue,
    builtin: builtinValue,
    healthUrl: healthUrlValue as string | null,
    transport,
  };
}

/**
 * Update or create service configuration.
 */
export function updateServiceConfig(
  serviceId: string,
  body: Record<string, unknown>,
  options?: { createIfMissing?: boolean },
): McpServiceConfig {
  const createIfMissing = options?.createIfMissing ?? false;
  const current = MCP_SERVICE_CONFIGS[serviceId];

  if (current === undefined && !createIfMissing) {
    throw new Error(`Unknown service: ${serviceId}`);
  }

  if (current !== undefined) {
    if (current.builtin && body.path !== undefined) {
      throw new Error('builtin service path is immutable');
    }
    if (current.builtin && body.builtin === false) {
      throw new Error('builtin service cannot be downgraded');
    }
  }

  const payload = normalizeServiceConfigPayload(serviceId, body);

  let updated: McpServiceConfig;
  if (current === undefined) {
    updated = payload;
  } else {
    updated = {
      serviceId: current.serviceId,
      name: payload.name,
      path: current.builtin ? current.path : payload.path,
      enabled: payload.enabled,
      builtin: current.builtin,
      healthUrl: payload.healthUrl,
      transport: payload.transport,
    };
  }

  MCP_SERVICE_CONFIGS[updated.serviceId] = updated;
  if (!(updated.serviceId in MCP_SERVICE_HEALTH_CACHE)) {
    MCP_SERVICE_HEALTH_CACHE[updated.serviceId] = 'unknown';
  }
  return updated;
}

/**
 * Enable or disable a service.
 */
export function setServiceEnabled(serviceId: string, enabled: boolean): McpServiceConfig {
  const config = MCP_SERVICE_CONFIGS[serviceId];
  if (config === undefined) {
    throw new Error(`Unknown service: ${serviceId}`);
  }
  if (config.builtin && !enabled) {
    throw new Error('builtin service cannot be disabled');
  }

  const updated: McpServiceConfig = {
    ...config,
    enabled,
  };
  MCP_SERVICE_CONFIGS[serviceId] = updated;
  return updated;
}

/**
 * Refresh the service health status cache.
 */
export function refreshServiceHealthCache(services: Record<string, string>): void {
  for (const serviceId of Object.keys(MCP_SERVICE_CONFIGS)) {
    MCP_SERVICE_HEALTH_CACHE[serviceId] = serviceRuntimeStatus(serviceId, services);
  }
}
